ark/phenotyping/som_utils.py

```diff
@@ -55,6 +56,75 @@ def normalize_rows(pixel_data, channels, include_seg_label=True):
+
+    Args:
+        tiff_dir (str):
+            Name of the directory containing the tiff files
+        test_fov (str):
+            example fov used to check channel names
+        img_sub_folder (str):
+            sub-folder within each FOV containing image data
+        channels (list): list of channels to use for analysis"""
+
+    # convert to path-compatible format
+    if img_sub_folder is None:
+        img_sub_folder = ''
+
+    # get all channels within example FOV
+    all_channels = io_utils.list_files(os.path.join(tiff_dir, test_fov, img_sub_folder))
+
+    # loop over each user-provided channel
+    for channel in channels:
+        # check for substring matching
+        matches = [match for match in all_channels if channel in match]
+
+        if len(matches) > 1:
+            warnings.warn('You selected {} as the channel to analyze, but there were multiple '
+                          'close matches: {}. Make sure you selected the correct version of the '
+                          'channel for inclusion in clustering'.format(channel, matches))
+
+
+def smooth_channels(fovs, tiff_dir, img_sub_folder, channels, smooth_vals):
+    """Adds additional smoothing for selected channels as a preprocessing step
+    Args:
+        fovs (list):
+            List of fovs to process
+        tiff_dir (str):
+            Name of the directory containing the tiff files
+        img_sub_folder (str): sub-folder within each FOV containing image data
+        channels (list): list of channels to apply smoothing to
+        smooth_vals (list or int): amount to smooth channels. If a single int, applies
+            to all channels. Otherwise, a custom value per channel can be supplied"""
+
+    # no output if no channels specified
+    if channels is None or len(channels) == 0:
+        return
+
+    # convert to path-compatible format
+    if img_sub_folder is None:
+        img_sub_folder = ''
+
+    # convert int to list of same length
+    if type(smooth_vals) is int:
+        smooth_vals = [smooth_vals for _ in range(len(channels))]
+    elif type(smooth_vals) is list:
+        if len(smooth_vals) != len(channels):
+            raise ValueError("A list was provided for variable smooth_vals, but it does not "
+                             "have the same length as the list of channels provided")
+    else:
+        raise ValueError("Variable smooth_vals must be either a single integer or a list")
+
+    for fov in fovs:
+        for idx, chan in enumerate(channels):
+            img = load_utils.load_imgs_from_tree(data_dir=tiff_dir, img_sub_folder=img_sub_folder,
+                                                 fovs=[fov], channels=[chan]).values[0, :, :, 0]
+            chan_out = ndimage.gaussian_filter(img, sigma=smooth_vals[idx])
+            imsave(os.path.join(tiff_dir, fov, img_sub_folder, chan + '_smoothed.tiff'),
+                   chan_out, check_contrast=False)
```

Comment:
On average how many channels will need additional smoothing in a cohort? If it's frequently going to be a lot, might be more efficient to put this in `create_fov_pixel_data` (which is enclosed in the per-FOV loop in `create_pixel_matrix`) to prevent the inefficiency of looping over each `channel` once to add the first `gaussian_filter`, then apply the "standard" `gaussian_filter` again in `create_fov_pixel_data`.

If only a few channels will need additional smoothing at most, this is a moot point.

Reply:
In my opinion, it will usually be a small number of markers that need additional smoothing.